File: core/src/main/java/nl/t64/game/rpg/screens/world/conversation/ConversationDialogListener.java
package nl.t64.game.rpg.screens.world.conversation;

import com.badlogic.gdx.Input;
import com.badlogic.gdx.scenes.scene2d.InputEvent;
import com.badlogic.gdx.scenes.scene2d.InputListener;
import com.badlogic.gdx.scenes.scene2d.ui.List;
import nl.t64.game.rpg.conversation.ConversationChoice;


class ConversationDialogListener extends InputListener {

    private final List<ConversationChoice> answers;
    private final Runnable selectAnswers;

    ConversationDialogListener(List<ConversationChoice> answers, Runnable selectAnser) {
        this.answers = answers;
        this.selectAnswers = selectAnser;
    }

    @Override
    public boolean keyDown(InputEvent event, int keycode) {
        switch (keycode) {
            case Input.Keys.UP -> inputUp();
            case Input.Keys.DOWN -> inputDown();
            case Input.Keys.ENTER -> inputEnter();
        }
        return true;
    }

    private void inputUp() {
        if (answers.getSelectedIndex() > 0) {
            answers.setSelectedIndex(answers.getSelectedIndex() - 1);
        }
    }

    private void inputDown() {
        if (answers.getSelectedIndex() < answers.getItems().size - 1)
            answers.setSelectedIndex(answers.getSelectedIndex() + 1);
    }

    private void inputEnter() {
        selectAnswers.run();
    }

}

File: core/src/nl/t64/game/rpg/screens/world/conversation/ConversationDialogListener.kt
package nl.t64.game.rpg.screens.world.conversation

import com.badlogic.gdx.Input
import com.badlogic.gdx.scenes.scene2d.InputEvent
import com.badlogic.gdx.scenes.scene2d.InputListener
import com.badlogic.gdx.scenes.scene2d.ui.List
import nl.t64.game.rpg.Utils.audioManager
import nl.t64.game.rpg.audio.AudioCommand
import nl.t64.game.rpg.audio.AudioEvent
import nl.t64.game.rpg.components.conversation.ConversationChoice
import nl.t64.game.rpg.constants.Constant


internal class ConversationDialogListener(
    private val answers: List<ConversationChoice>,
    private val selectAnswer: () -> Unit
) : InputListener() {

    override fun keyDown(event: InputEvent, keycode: Int): Boolean {
        when (keycode) {
            Input.Keys.UP -> inputUp()
            Input.Keys.DOWN -> inputDown()
            Constant.KEYCODE_BOTTOM, Input.Keys.ENTER, Input.Keys.A -> inputConfirm()
        }
        return true
    }

    private fun inputUp() {
        if (answers.items.size > 1 && answers.selectedIndex > 0) {
            audioManager.handle(AudioCommand.SE_PLAY_ONCE, AudioEvent.SE_CONVERSATION_CURSOR)
            answers.selectedIndex = answers.selectedIndex - 1
        }
        if (answers.selectedIndex == -1) {
            audioManager.handle(AudioCommand.SE_PLAY_ONCE, AudioEvent.SE_CONVERSATION_CURSOR)
            answers.selectedIndex = answers.items.size - 1
        }
    }

    private fun inputDown() {
        if (answers.items.size > 1 && answers.selectedIndex < answers.items.size - 1) {
            audioManager.handle(AudioCommand.SE_PLAY_ONCE, AudioEvent.SE_CONVERSATION_CURSOR)
            answers.selectedIndex = answers.selectedIndex + 1
        }
    }

    private fun inputConfirm() {
        if (answers.selectedIndex != -1) {
            selectAnswer.invoke()
        }
    }

}
